pydantic_ai_slim/pydantic_ai/mcp.py

```diff
@@ -50,6 +59,166 @@
 )
 
 
+def _is_mcp_run_python_server(toolset: AbstractToolset[Any]) -> bool:
+    """Check if a toolset is an mcp-run-python server to avoid deadlock."""
+    command = getattr(toolset, 'command', None)
+    args = getattr(toolset, 'args', None)
+    if command is not None and args is not None:
+        command_str = ' '.join([str(command)] + [str(arg) for arg in args])
+        return 'mcp-run-python' in command_str
+    return False
+
+
+class _FilteredToolset(AbstractToolset[Any]):
+    """Toolset wrapper that excludes mcp-run-python servers to prevent deadlock."""
+
+    def __init__(self, wrapped: AbstractToolset[Any]):
+        from .toolsets.combined import CombinedToolset
+
+        self.wrapped = wrapped
+
+        # If it's a CombinedToolset, filter out mcp-run-python toolsets before any get_tools() calls
+        if isinstance(wrapped, CombinedToolset):
```

Comment:
This won't work if there are multiple levels of nested toolsets, unfortunately.

Could we require users to specify their own toolset to provide to mcp_run_python, and make them responsible for not including mcp_run_python itself? Maybe there's another point where we could detect this and raise an error, instead of trying to filter proactively?